src/main/java/redis/clients/jedis/JedisSentinelPool.java

```diff
@@ -315,7 +317,12 @@ public void onMessage(String channel, String message) {
 
         } catch (JedisException e) {
 
-          if (running.get()) {
+          try {
+            lock.unlock();
```

Comment:
In this case we'll always get to:
line 341: log.debug("lock is not held by MasterListener");

Reply:
I think we need to try release lock at both the places.
Here at line 321, we need to release because after this we sleep for 5 seconds. I don't want anyone trying to shutdown, wait for it. 

The reason we need to shutdown in finally (line 339) is if there is some unchecked exception which we are not catching, it can result in lock being never released. In that case anyone calling shutdown will be blocked for ever